raven/middleware.py

```diff
@@ -9,6 +9,37 @@
 
 from raven.utils.wsgi import (
     get_current_url, get_headers, get_environ)
+from raven.utils.six import Iterator, next
+
+
+class ClosingIterator(Iterator):
+    """
+    An iterator that is implements a ``close`` method as-per
+    WSGI recommendation.
+    """
+    def __init__(self, sentry, iterable, environ):
+        self.sentry = sentry
+        self.environ = environ
+        self.iterable = iter(iterable)
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        try:
+            return next(self.iterable)
+        except StopIteration:
+            # propagate up the normal StopIteration
+            raise
+        except Exception:
+            # but capture any other exception, then re-raise
+            self.sentry.handle_exception(self.environ)
+            raise
+
+    def close(self):
+        self.sentry.client.context.clear()
```

Comment:
This is then working under the assumption below, that it's the consumer's job to `close` the iterator, which makes sense then, and will produce undesired results if the consumer doesn't close it. But that seems par for the course in expectations of this method existing.